dev/usb/class/cdcserial/cdcserial.c

```diff
@@ -0,0 +1,331 @@
+    W(0x09),        /* Total Length (incl ifc, ept) */
+    0x00,           /* # Interfaces */
+    0x01,           /* Cfg Value */
+    0x00,           /* Cfg String */
+    0xc0,           /* Attributes -- self powered */
+    250,            /* Power Consumption - 500mA */
+};
+
+static const uint8_t ctrl_if_descriptor[] = {
+    0x09,           /* length */
+    INTERFACE,      /* type */
+    0x00,           /* interface num */
+    0x00,           /* alternates */
+    0x01,           /* endpoint count */
+    0x02,           /* interface class */
+    0x02,           /* interface subclass */
+    0x01,           /* interface protocol */
+    0x00,           /* string index */
+
+    /*Call Management Functional Descriptor*/
+    0x05,   /* bFunctionLength */
+    0x24,   /* bDescriptorType: CS_INTERFACE */
+    0x01,   /* bDescriptorSubtype: Call Management Func Desc */
+    0x00,   /* bmCapabilities: D0+D1 */
+    0x00,   /* bDataInterface: 1 */
+
+    /*ACM Functional Descriptor*/
+    0x04,   /* bFunctionLength */
+    0x24,   /* bDescriptorType: CS_INTERFACE */
+    0x02,   /* bDescriptorSubtype: Abstract Control Management desc */
+    0x02,   /* bmCapabilities */
+
+    /*Union Functional Descriptor*/
+    0x05,   /* bFunctionLength */
+    0x24,   /* bDescriptorType: CS_INTERFACE */
+    0x06,   /* bDescriptorSubtype: Union func desc */
+    0x00,   /* bMasterInterface: Communication class interface */
+    0x01,   /* bSlaveInterface0: Data Class Interface */
+
+    /* endpoint 1 IN */
+    0x07,           /* length */
+    ENDPOINT,       /* type */
+    0x82,           /* address: 1 IN */
+    0x03,           /* type: bulk */
+    W(8),           /* max packet size: 64 */
+    0xFF,           /* interval */
+};
+
+static const uint8_t data_if_descriptor[] = {
+    /*Data class interface descriptor*/
+    0x09,       /* bLength: Endpoint Descriptor size */
+    INTERFACE,  /* bDescriptorType: */
+    0x01,       /* bInterfaceNumber: Number of Interface */
+    0x00,       /* bAlternateSetting: Alternate setting */
+    0x02,       /* bNumEndpoints: Two endpoints used */
+    0x0A,       /* bInterfaceClass: CDC */
+    0x00,       /* bInterfaceSubClass: */
+    0x00,       /* bInterfaceProtocol: */
+    0x00,       /* iInterface: */
+
+    /*Endpoint OUT Descriptor*/
+    0x07,      /* bLength: Endpoint Descriptor size */
+    ENDPOINT,  /* bDescriptorType: Endpoint */
+    0x01,      /* bEndpointAddress */
+    0x02,      /* bmAttributes: Bulk */
+    0x40,      /* wMaxPacketSize: */
+    0x00,
+    0x00,      /* bInterval: ignore for Bulk transfer */
+
+    /*Endpoint IN Descriptor*/
+    0x07,      /* bLength: Endpoint Descriptor size */
+    ENDPOINT,  /* bDescriptorType: Endpoint */
+    0x81,      /* bEndpointAddress */
+    0x02,      /* bmAttributes: Bulk */
+    0x40,      /* wMaxPacketSize: */
+    0x00,
+    0x00       /* bInterval */
+};
+
+static const uchar langid[] = { 0x04, 0x03, 0x09, 0x04 };
```

Comment:
I forget actually. It's something to do with the encoding of any string descriptors, and this is the generic one. But as I pointed out this particular part is part of the overall usb config and actually doesn't belong here anyway, it's up to the target/project to define the overall device descriptor. This code should just append some interface descriptor to the list.